File: LinkedList/Lecture 22/LRU Cache.cpp
/* In this question, we are asked to design LRU (Least Recently Used) data structure. 
To solve the question, we will require Two Unordered Maps:
        a) One will store all the Key Value pairs
        b) One will store the iterators corresponding to all the Keys stored in First Map
        
We need to know always which data is recently used or not. For that purpose, we will use Doubly LinkedLists, that is, Lists in C++ STL .

Our LRU-Cache will have the following functions:
a) LRUCache(int capacity) --> Constructor to initialise the Max Capacity of our Cache
b) int get(int key) --> Returns the value assoicated with the key, if present, else returns -1
c) void put(int key , int value) --> Update the value associated with the key, if present. Otherwise, make the new entry. And if Capacity exceeds then delete the Least Recently Used(Last Node) and then add the key-value pair
*/


class LRUCache {
public:
    
    // Map M will store all the Key-Value Pairs 
    
    unordered_map<int , int> m ;
    
    // Map Address will help us to get the iterator(pointer) of a particular Node present in our Doubly Linked List
    
    unordered_map<int , list<int> :: iterator> address ;
    
    // Doubly LinkedList L will help us to keep track of the recently used data and also to delete as well as make new entiries in our Map M
    
    list<int> l ;
    
    // Cap will store the maximum Capacity of our LRU Data Structure
    
    int cap ;
    
    // Siz will store the Current Size of our LRU Data Structure
    
    int siz  ;
    
    // Initally, Cap is intialised with Capacity(given by the user) and Siz as 0
    
    LRUCache(int capacity) 
    {
        cap = capacity ;
        
        siz = 0 ;
    }
    
    // This function will return the value associated with a particular Key and make the Key move to the front of our Doubly LinkedList
    
    int get(int key) 
    {
        // If the Key is not present, simply return -1
        
        if(m.find(key) == m.end() )
            return -1 ;
        
        // Store the Address of the Key using Address Map
        
        list<int> :: iterator it = address[key] ;
        
        // Erase the Key from the List and the Address Map
        
        l.erase(it) ;
        address.erase(key) ;
        
        // Put the Key again at the Front of our Doubly LinkedList & insert in the Address Map again
        
        l.push_front(key) ;
        address[key] = l.begin() ;
        
        // Return the value associated with the Key using Map M
        
        return m[key] ;
        
    }
    
    // This function inserts the Key-Value pair if not present. Otherwise update the Key-Value pair with the new Value
    
    void put(int key, int value) 
    {
        // If Key is already present in Map M, delete it & decrement Siz by 1
        
        if(m.find(key) != m.end() )
        {
            l.erase(address[key]) ;
            
            address.erase(key) ;
            
            m.erase(key) ;
            
            siz-- ;
        }
        
        // If Siz is equal to Cap, that is, Capacity is full, we need to delete the Last Node from our Doubly LinkedList
        
        if(siz == cap)
        {
            // Store the Last Node's Value in variable K
            
            int k = l.back() ;
            
            // Delete the Node from Doubly LinkedList 
            
            l.pop_back() ;
            
            // Remove K from Address Map & M Map as well
            
            address.erase(k) ;
            m.erase(k) ;
            
            // Decrement Siz by 1
            
            siz-- ;
        }
        
        // Now, we can insert the Key-Value Pair into our Map M as well as our List L
        
        siz++ ;
        
        // Make the Key as the First Node of our Doubly LinkedList L
        
        l.push_front(key) ;
        
        // Store the iterator / pointer of Key into Address Map
        
        address[key] = l.begin() ;
        
        // Store the Key Value pair in Map M
        
        m[key] = value ;
    }
};


/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

/*
Time Complexity:  O(1) {Average Case Time Complexity}
Space Complexity: O(N)
*/

File: LinkedList/Lecture 22/LRU Cache.java
//  We will make a double linked list - in which we will have a dummy head and tail for traversals from both side. So the nodes apperaing after head - would be the one that are recently used, and at the time whene size is exceeded we will remove node form the back side.

// We will be using hashmap, so that when get function is called, we will search the element in our doubly linked list

class LRUCache {
    
    Node head = new Node(0,0);  // for key-value
    Node tail = new Node(0,0);
    
    HashMap<Integer, Node> map = new HashMap<>();   // integer - key, Node - where there is our key
    
    int size;
    
    // intially lets point head next to tail, and it's prev to head

    public LRUCache(int capacity) 
    {
        size = capacity;
        
        head.next = tail;
        tail.prev = head;
    }
    
    public int get(int key) 
    {
        // to get a key - we will first search it our hashmap 
        // if found: 
        //      --> with the help of key-value - we will search that node
        //      --> remove that node, and place it in the fornt - as it is now recently used
        //      --> remove it from hashmap
        //      --> update the hashmap with the new node address
        // if not found:
        //      --> return -1
        
        if(map.containsKey(key) == true)
        {
            // found
            
            // store the node value from the hashmap
            Node temp = map.get(key);
            
            // remove this node
            remove(temp);
            
            // insert it now
            insert(temp);
            
            // now our insert and remove function - automatically changes our hashmap values too
            return temp.value;
        }
        else
        {
            return -1;
        }
    }
    
    public void put(int key, int value) 
    {
        // now to put the key-value pair in our LRU cache we will have to keep track of our size
        
        // if the size of hashmap is lesser than the size allotted - we will insert it
        // if the size is equal, we will remove the element from last and insert it again
        
        // if map already contains the value we are abt to insert - then again remove it first and then insert
        
        if(map.containsKey(key) == true)
        {
            remove(map.get(key));
        }
        
        if(map.size() == size)
        {
            remove(tail.prev);  // reomve from back
        }
        
        Node temp = new Node(key, value);
        // insert it 
        insert(temp);
    }
    
    class Node
    {
        // In this question we will be taking help of doubly linked list
        int value, key;
        Node next;
        Node prev;
    
        // these key-value pairs are used, as further we will be using hashmap too, so that the get(key) method can be implemented in O(1) time - as it is a search operation
        Node(int key1, int value1)
        {
            key = key1;
            value = value1;
        }
    }

    // now we will make our basic - insert and delete operation
    private void insert(Node temp)
    {
        // always our insertion will happen from the head side - so as to denote that this is the most recently used one
    
        // before adding in our linked list - we will add this element in our hashmap too
        map.put(temp.key, temp);
    
        // now we will make the adjustments
        temp.next = head.next;
        head.next.prev = temp;
        head.next = temp;
        temp.prev = head;
    }

    private void remove(Node temp)
    {
        // to remove, we will delete from the tail side
    
        // first we will remove it from the hashmap
        map.remove(temp.key);
    
        // now from the list
        temp.prev.next = temp.next;
        temp.next.prev = temp.prev;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */ 